frontend/src/metabase/lib/expressions/typechecker.js

```diff
@@ -82,3 +82,41 @@ export function typeCheck(cst, rootType) {
   }
   new TypeChecker().visit(cst);
 }
+
+export function compactSyntaxTree(node) {
+  if (!node) {
+    return;
+  }
+  let result = node;
+  const children = node.children;
+  switch (node.name) {
+    case "expression":
+    case "atomicExpression":
+    case "boolean":
+    case "booleanExpression":
+    case "booleanUnaryExpression":
+      children.expression = children.expression.map(n => {
```

Comment:
This is a horrendous mutation. I'm working on an immutable flavor of this function.